docs/custom_validators.md

```diff
@@ -0,0 +1,158 @@
+# Implementing custom validators
+
+With custom validators, you can implement business logic in Python. Schema-enforcer will automatically
+load your plugins from the `validator_directory` and run them against your host data.
+
+The validator plugin provides two base classes: ModelValidation and JmesPathModelValidation. The former can be used
+when you want to implement all logic and the latter can be used as a shortcut for jmespath validation.
+
+## BaseValidation
+
+Use this class to implement arbitrary validation logic in Python. In order to work correctly, your Python script must meet
+the following criteria:
+
+1. Exist in the `validator_directory` dir.
+2. Include a subclass of the BaseValidation class to correctly register with schema-enforcer.
+3. Ensure you call `super().__init__()` in your class `__init__` if you override.
+4. Provide a class method in your subclass with the following signature:
+`def validate(data: dict, strict: bool):`
+
+   * Data is a dictionary of variables on a per-host basis.
+   * Strict is set to true when the strict flag is set via the CLI. You can use this to offer strict validation behavior
+   or ignore it if not needed.
+
+The name of your class will be used as the schema-id for mapping purposes. You can override the default schema ID
+by providing a class-level `id` variable.
+
+Helper functions are provided to add pass/fail results:
+
+```
+def add_validation_error(self, message: str, **kwargs):
+    """Add validator error to results.
+    Args:
+      message (str): error message
+      kwargs (optional): additional arguments to add to ValidationResult when required
+    """
+
+def add_validation_pass(self, **kwargs):
+    """Add validator pass to results.
+    Args:
+      kwargs (optional): additional arguments to add to ValidationResult when required
+    """
+```
+In most cases, you will not need to provide kwargs. However, if you find a use case that requires updating other fields
+in the ValidationResult, you can send the key/value pairs to update the result directly. This is for advanced users only.
+
+## JmesPathModelValidation
+
+Use this class for basic validation using [jmespath](https://jmespath.org/) expressions to query specific values in your data. In order to work correctly, your Python script must meet
+the following criteria:
+
+1. Exist in the `validator_directory` dir.
+2. Include a subclass of the JmesPathModelValidation class to correctly register with schema-enforcer.
+3. Provide the following class level variables:
+
+   * `top_level_properties`: Field for mapping of validator to data
+   * `id`: Schema ID to use for reporting purposes (optional - defaults to class name)
+   * `left`: Jmespath expression to query your host data
```

Comment:
I am the one who indicated left/right, but I did so out of lack of imagination. Just want to make sure we are all good with the name here.

Reply:
Another option is lhs (left-hand side) and rhs (right-hand side).

Reply:
left / right sounds good to me